### PR TITLE
Add rate limiting

### PR DESCRIPTION
**Pull request by sarahsturgeon**

### CODE REVIEW

File: lua/entities/gmod_wire_expression2/core/custom/streamcore.lua [hunk 1]
```diff
@@ -5,24 +5,77 @@ CreateConVar( "streamcore_antispam_ignoreadmins", 1, FCVAR_SERVER_CAN_EXECUTE )
...
+        streamVolume = 0.25,
+        streamStop = 0.25
+    }
+}
+
+local function setLastUse( chip, funcName )
+    local ent = chip.entity
+
+    local throttles = ent.streamcoreThrottle
+    if not throttles then ent.streamcoreThrottle = {} end
```
Comment (samuelWilliams99) [inline lua/entities/gmod_wire_expression2/core/custom/streamcore.lua]: gotta assign throttles to this new table else it'll still be nil when you set it to CurTime
```suggestion
    if not throttles then
        throttles = {}
        ent.streamcoreThrottle = throttles
    end
```

File: lua/entities/gmod_wire_expression2/core/custom/streamcore.lua [hunk 2]
```diff
@@ -5,24 +5,77 @@ CreateConVar( "streamcore_antispam_ignoreadmins", 1, FCVAR_SERVER_CAN_EXECUTE )
...
+
+local function isThrottled( chip, funcName )
+    local ply = chip.player
+    local ent = chip.entity
+
+    if ply:IsAdmin() then return false end
+
+    local throttles = ent.streamcoreThrottle
+    if not throttles then
+        ent.streamcoreThrottle = {}
```
Comment (samuelWilliams99) [inline lua/entities/gmod_wire_expression2/core/custom/streamcore.lua]: Same as before

File: lua/entities/gmod_wire_expression2/core/custom/streamcore.lua [hunk 3]
```diff
@@ -110,55 +168,80 @@ e2function void entity:streamStart( id, volume, string url )
...
+    if volume == streamtbl[2] then return end
+
+    streams[index][2] = volume
+
+    net.Start( "CFC_SC_StreamVolume" )
+        net.WriteString( index )
+        net.WriteFloat( volume )
+    net.Broadcast()
+
+    setLastUse(self, "streamVolume")
```
Comment (samuelWilliams99) [inline lua/entities/gmod_wire_expression2/core/custom/streamcore.lua]:
```suggestion
    setLastUse( self, "streamVolume" )
```

File: lua/entities/gmod_wire_expression2/core/custom/streamcore.lua [hunk 4]
```diff
@@ -110,55 +168,80 @@ e2function void entity:streamStart( id, volume, string url )
...
+
+    if radius == streamtbl[3] then return end
+
+    streams[index][3] = radius
+    net.Start( "CFC_SC_StreamRadius" )
+        net.WriteString( index )
+        net.WriteFloat( radius )
+    net.Broadcast()
+
+    setLastUse(self, "streamRadius")
```
Comment (samuelWilliams99) [inline lua/entities/gmod_wire_expression2/core/custom/streamcore.lua]:
```suggestion
    setLastUse( self, "streamRadius" )
```

**Review by samuelWilliams99 — CHANGES_REQUESTED** [review]

File: lua/entities/gmod_wire_expression2/core/custom/streamcore.lua [hunk 5]
```diff
@@ -5,24 +5,77 @@ CreateConVar( "streamcore_antispam_ignoreadmins", 1, FCVAR_SERVER_CAN_EXECUTE )
...
+util.AddNetworkString( "CFC_SC_StreamRadius" )
+util.AddNetworkString( "CFC_SC_StreamHelp" )
+
+local rawget = rawget
+local CurTime = CurTime
+local clamp = math.Clamp
+
+local stringLen = string.len
+local stringSub = string.sub
+local stringTrim = string.Trimp
```
Comment (sarahsturgeon) [inline lua/entities/gmod_wire_expression2/core/custom/streamcore.lua]:
```suggestion
local stringTrim = string.Trim
```

File: lua/entities/gmod_wire_expression2/core/custom/streamcore.lua (hunk 2 above)
Comment (sarahsturgeon) [inline lua/entities/gmod_wire_expression2/core/custom/streamcore.lua]:
```suggestion
        throttles = {}
        ent.streamcoreThrottle = throttles
```

**Review by samuelWilliams99 — DISMISSED** [review]

**Comment by sarahsturgeon** [comment]
Tested, works as expected 👍 

Going to throw it on the server and let the players bug test it :^)

**Pull request merged**